lib/history/sqlite.go

```diff
@@ -0,0 +1,212 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package history
+
+import (
+	"database/sql"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/gravitational/trace"
+	_ "github.com/mattn/go-sqlite3" // initialize sqlite3
+	log "github.com/sirupsen/logrus"
+)
+
+// SQLiteTimeline represents a timeline of cluster status events. The timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// size specifies the max size of the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sql.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *Cluster
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specifies size.
+func NewSQLiteTimeline(database *sql.DB, size int) (Timeline, error) {
+	if err := initEventTable(database); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		size:       size,
+		database:   database,
+		lastStatus: &Cluster{},
+	}, nil
+}
+
+// initEventTable initializes event table schema.
+func initEventTable(db *sql.DB) error {
+	createTables, err := db.Prepare(createTableEvent)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if _, err := createTables.Exec(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// RecordStatus records differences of the previous status to the provided
+// status into the Timeline.
+func (t *SQLiteTimeline) RecordStatus(status *pb.SystemStatus) {
+	cluster := parseSystemStatus(status)
+	events := t.lastStatus.diffCluster(cluster)
+	if len(events) == 0 {
+		return
+	}
+
+	for _, event := range events {
+		t.addEvent(event)
```

Comment:
This will run each `INSERT` in its own transaction which is sub-optimal - I'd consider grouping these within a single transaction.